Thesis.org

```diff
@@ -1,11 +1,725 @@
+#+begin_src clojure -n 1
+  (defn then [prog cont]
+    {:pre [(or (bind? prog) (prog? prog) (c/item? prog))
+	   (fn? cont)]
+     :post [(bind? %)]}
+    (if (bind? prog)
+      (make-bind (bind-item prog) (fn [x] (then ((bind-continuation prog) x) cont)))
+      (make-bind (if (c/item? prog) (make-prog prog) prog) cont)))
+#+end_src
+To give the developer feedback in case they make an error, we add ~:pre~ and ~:post~ annotations, which let ClojureScript know to check the types that come into and out of our function. In this case we say that the ~prog~ can be a ~Prog~, ~Bind~ or even an ~Item~. The ~cont~ parameter needs to be a function and the result of our operation should always return a ~Bind~.
+** TODO runner
+The place for our monad to be executed in is the ~runner~. It will receive a ~Prog~ or ~Bind~. The ~runner~ is the most complex function in our API because of all the things it needs to do:
+1. Show current step
+2. Bookkeeping of state for steps
+3. Catch emitted ~Commits~
+4. Make sure implicit state is passed to the ~Prog~ (without leaking own state)
+5. And optimize function calls to prevent stackoverflow
+As such we will show the code in its entirety once in listing 21 and go in depth about individual parts one after another.
+#+CAPTION: The entire definition of the ~runner~ function using trampolines, state-management, lenses and actions.
+#+begin_src clojure -n 1
+  (defn runner [b]
+    {:pre [(or (bind? b) (prog? b))]}
+    (core/local-state
+     b
+     (core/dynamic
+      (fn [[_ inner]]
+	(core/handle-action
+	 (core/focus
+	  first-lens
+	  (show inner))
+	 (fn [[outter st] ac]
+	   (if (and (commit? ac) (bind? st))
+	     (core/return :state [outter ((bind-continuation st) (commit-payload ac))])
+	     (core/return :action ac))))))))
+#+end_src
+*** Basic Functionality
+The goal of the ~runner~ is to hold and display what is inside of a ~Bind~ (or ~Prog~). While it does this, it wraps the ~Bind~ and waits for a ~Commit~ which will trigger it to call the ~continuation~ of the ~Bind~.
+To understand this better we're going to focus on lines 7-14 of listing 21. First notice the ~core/handle-action~ call. This function takes two parameters. First an ~Item~ to display, second a ~function~ to call once the ~Item~ emits something. The function that will be called can be seen at the bottom on lines 11-14. We don't directly pass an ~Item~ however. We pass a ~focus~ function. This is done to ~core/focus~ the state on a specific part. Namely limiting the implicitly passed state to what is outside of the runner, instead of leaking the bookkeeping state of the runner downwards. This function again takes two parameters. First a lens (function of two arities) and second an ~Item~. We will talk more about the lense in the next chapter.
+The second parameter is the in our API defined ~show~ function, which just takes either ~Bind~, ~Prog~ or ~Item~ and unwraps it to an ~Item~ again.
```

Comment:
:D This sentence doesn't work in English. "The second parameter is the `show` function defined in our API."

Reply:
```suggestion
The second parameter is the ~show~ function defined in our API, which just takes either ~Bind~, ~Prog~ or ~Item~ and unwraps it to an ~Item~ again.
```

Ups, hehe.